python/pfs/utils/optimize_black_dots.py

```diff
@@ -0,0 +1,482 @@
+
+    def return_list_of_prepared_observations_and_predictions(self):
+        """Returns the observed and predicted positions
+        """
+        return self.list_of_prepared_observations_and_predictions
+
+    def predict_positions(self, mcs_data_all, type_of_run='theta'):
+        """Predict positions of spots which were not observed
+
+        Parameters
+        ----------
+        mcs_data_all: `np.array`
+           positions of observed spots
+         type_of_run: `str`
+             describing which kind of movement is being done
+             (theta or phi)
+
+        Returns
+        ----------
+        prepared_observations_and_predictions `list`
+            lists contaning 4 arrays, which contanin observed position in x and y,
+            and predicted position in x and y
+        """
+
+        # lists which will contain the positions which have been observed
+        # 2 runs and 2 coordinates = 4 lists
+        list_of_actual_position_x_observed = []
+        list_of_actual_position_y_observed = []
+
+        # lists which will contain the position which were not observed
+        # and we predicted with this algorithm
+        list_of_predicted_position_x_not_observed = []
+        list_of_predicted_position_y_not_observed = []
+
+        for i in range(0, 2394):
+            try:
+                x_measurments = mcs_data_all[0, i]
+                y_measurments = mcs_data_all[1, i]
+
+                # index of the points which are not observed
+                idx = np.isfinite(x_measurments) & np.isfinite(y_measurments)
+
+                # select the point that are more than 2 mm away from median of all of the points
+                # replace them with np.nan
+                # these points have been attributed to wrong cobras
+                # ignore warning if idx_1 is all False, i.e., you didnt observe any of the points
+                with warnings.catch_warnings():
+                    warnings.simplefilter("ignore", category=RuntimeWarning)
+                    large_outliers_selection = (np.abs(y_measurments -
+                                                       np.median(y_measurments[idx])) > 2) |\
+                        (np.abs(x_measurments - np.median(x_measurments[idx])) > 2)
+                y_measurments[large_outliers_selection] = np.nan
+                x_measurments[large_outliers_selection] = np.nan
+
+                # if np.sum(idx_1)<10:
+                #    raise ValueError('too few points')
+
+                # create prediction for where the points should be, when you do not see points
+                # if you do not see points from both sides of the black dots, do simpler extrapolation
+                # (np.sum(np.diff(idx_1)) == 1)
+                # if you see points from both sides of the black dots, do more complex interpolation
+                # (np.sum(np.diff(idx_1)) != 1)
```

Comment:
Remove the commented out code.

Reply:
I was trying to describe how I determine if the points are seen on both sides or only one. Reformatted comment so that is more clear what is happening.